deploy_to_s3.py

```diff
@@ -0,0 +1,123 @@
+        # font resources: one month cache
+        if 'font-awesome-4.7.0' in path or 'fonts' in path:
+            return 2592000
+        # minified JS: one week cache
+        if '.min.js' in filename or '.min.css' in filename:
+            return 604800
+        # one day cache
+        return 86400
+    # default 30min cache for html pages
+    return 1800
+
+def get_cache_control(path, filename, production=False):
+    if not production:
+        # one minute cache
+        return 'max-age=60'
+    return f'max-age={get_max_age(path, filename)}'
+
+
+def get_content_type(local_path):
+    mime = MimeTypes()
+    guessed_type = mime.guess_type(local_path)[0]
+    if guessed_type is None:
+        if local_path.endswith('.css.map') or local_path.endswith('.js.map'):
+            return 'application/json'
+        return 'text/html'
+    return guessed_type
+
+if __name__ == '__main__':
+    args = docopt(__doc__)
+    bucket_name = args['<bucket_name>']
+    bucket_prefix = args['<prefix>'] or '/'
+    production = args['--production']
+    site_dir = './_site'
+
+    if bucket_prefix[0] != '/':
+        bucket_prefix = '/' + bucket_prefix
+    print('Deploying site to ', f's3://{bucket_name}{bucket_prefix}')
+
+    # check site build exists
+    if not os.path.isdir(site_dir):
+        print('_site directory does not exist. You need to build the site before uploading')
+        exit(1)
+
+    # connect s3
+    s3_client = boto3.client('s3')
+
+    # list existing bucket contents
+    existing_keys = set(iterate_bucket(s3_client, bucket_name, bucket_prefix[1:]))
+    print('Bucket contains', len(existing_keys), 'pages')
+    uploaded = 0
+    redirected = 0
+
+    def upload_file_to_s3(path, filename):
+        # path to local file
+        local_path = os.path.join(path, filename)
+        # relative folder of file
+        s3_suffix = path[len(site_dir) + 1:]
+        # full path from root of bucket
+        s3_path = os.path.join(bucket_prefix, s3_suffix, filename)[1:]
+
+        # metadata to set on file
+        cache_control = get_cache_control(s3_suffix, filename, production=production)
+        content_type = get_content_type(local_path)
+
+        print('put', local_path, s3_path)
+        with open(local_path, 'rb') as fp:
+            s3_client.put_object(Bucket=bucket_name, Key=s3_path, Body=fp,
+                                 CacheControl=cache_control,
+                                 ContentType=content_type)
+
+        # setup redirects
+        html_path = f'{s3_path}.html'
+        if html_path in existing_keys:
+            print(f'redirect {html_path} to /{s3_path}')
+            s3_client.put_object(Bucket=bucket_name, Key=html_path,
+                                 WebsiteRedirectLocation=f'/{s3_path}')
+            # upload + redirect
+            return True, True
+        else:
+            # upload, no redirect
+            return True, False
+
+    with ThreadPoolExecutor(max_workers=5) as executor:
```

Comment:
Alternatively if we introduce an extra helper function:
```python
def walk_site_directory(site_dir):
    for (dirpath, dirnames, filenames) in os.walk(site_dir):
        for filename in filenames:
            yield dirpath, filename
```

The upload could be re-written as:
```python
with ThreadPoolExecutor(max_workers=5) as executor:
    uploads = executor.map(
        upload_file_to_s3(dirpath, filename)
        for (dirpath, filename) in walk_site_directory(site_dir)
    )

    uploaded, redirected = map(sum, (zip(*uploads)))
```

It might better use all the threads since it does not upload directory per directory.

Reply:
Also, is there a reason the number of threads is `5`?